Start hooking in execution environments

jbradberry opened the pull request:
##### SUMMARY

- add a user-configurable setting for the global default EE
- add the ability to set Project-level default execution environments
- add an execution environment admin role that Org admins can delegate
- add a new Container Registry managed credential type

related #7064 

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
 - API

##### AWX VERSION
```
awx: 14.1.0
```

This PR has been closed, and the changes ported over to https://github.com/shanemcd/awx/pull/19

jbradberry committed:
Add a new ExecutionEnvironment model

jbradberry committed:
Create the RBAC access class for execution environments

jbradberry committed:
Add in the basic list and detail api views

jbradberry committed:
Change OrganizationSerializer to show and set default_environment

jbradberry committed:
Add execution_environment and pull to the fields for UJs and UJTs

jbradberry committed:
Expose an API view for all of the execution environments under an org

jbradberry committed:
Expose execution environments in awxkit and awx-cli

jbradberry committed:
Add a new Swagger topic

jbradberry committed:
Make changes to support capture by the activity stream

Including exposing a new API view for a particular EE's activity
stream objects.

jbradberry committed:
Add a sublist api view for the UJTs that use a given execution environment

shanemcd committed:
Initial EE integration

jbradberry committed:
Run a receptor node in the dev environment

nixocio committed:
Add execution environments files (#7909)

Update navigation bar and routing system to add execution environments.
Also, add stub files for the remaining related work.

See: https://github.com/ansible/awx/issues/7885
Also: https://github.com/ansible/awx/issues/7884

AlanCoding committed:
Add Z to volume mount

Update to AWX execution environment
  use the special 2.9 container image
  revert setting back for merge

Fix another permission error by mapping 2 folders
  also create folders before running

AlanCoding committed:
Add Z to volume mount

Set ansible-runner back to main fork due to merge

AlanCoding committed:
Purge environment variables to work with ansible-runner changes

Remove inventory scripts show because they no longer exist

Remove reference to non-existent callback directory

Remove more references to removed paths

AlanCoding committed:
Respect user proot show paths when using containers

AlanCoding committed:
Fix pathing issue with custom credentials

also fix some minor flake8 issues

AlanCoding committed:
Fix pathing issue for credential file references

AlanCoding committed:
Fix project folder deletion

Fix another absolute path reference in containers

AlanCoding committed:
Deal with missing HOME env var

AlanCoding committed:
Initial minimal hooking up of JT EEs to jobs

jbradberry committed:
Add in more model changes around execution environments

- add a user-configurable setting for the global default EE
- add the ability to set Project-level default execution environments
- add an execution environment admin role that Org admins can delegate
- add a new Container Registry managed credential type

AlanCoding commented:
whoops, looks like there was a rebasing issue

AlanCoding commented on awx/main/conf.py:
```diff
@@ -164,6 +165,18 @@ def _load_default_license_from_file():
+register(
+    'DEFAULT_EXECUTION_ENVIRONMENT',
+    field_class=fields.PrimaryKeyRelatedField,
+    allow_null=True,
```
What would be the behavior when this is null?

jbradberry replied:
The most recent `managed_by_tower = True` version would be used.

AlanCoding replied:
So then we still need some code to pre-create a default EE? Or is that already in the branch somewhere?

jbradberry replied:
Yes, we still need it and it isn't in-place yet.

AlanCoding replied:
What if, instead of creating it in a migration of create_preload_data (I do not believe the latter is reliable), we create the default execution environment ad-hoc if a job tries to launch while no `managed_by_tower` execution environments exist? Just thinking that might be easier and more reliable.

jbradberry committed:
Choose which execution environment needs to be used when spinning up a job

and attach it.

jbradberry committed:
Enable Project.default_environment in the serializer

nixocio committed:
Add list Execution Environments (#8148)

Add list execution environments.

See: https://github.com/ansible/awx/issues/7886

AlanCoding committed:
Changes to get execution environments factories working (#8126)

nixocio committed:
Add feature to Add/Edit Execution Environments (#8165)

* Add feature to Add/Edit Execution Environments

Add feature to Add/Edit Execution Environments.

Also, add key for `ExecutionEnvironmentsList`.

See: https://github.com/ansible/awx/issues/7887

* Update registry credential label

nixocio committed:
Add details page for Execution Environments (#8172)

* Add feature to Add/Edit Execution Environments

Add feature to Add/Edit Execution Environments.

Also, add key for `ExecutionEnvironmentsList`.

See: https://github.com/ansible/awx/issues/7887

* Add details page for execution environments

Add details page for execution environments

See: https://github.com/ansible/awx/issues/8171

jbradberry committed:
Expose the user capabilities dict for EEs (#8208)

jbradberry committed:
Add a query optimization to ExecutionEnvironmentAccess

nixocio committed:
Update usage of summary_fields for execution environments (#8217)

Update usage of summary_fields for execution environments. Also, update
unit-tests to cover this change.

See: https://github.com/ansible/awx/issues/8216

AlanCoding committed:
Collections volume permission fix, and container group fix

Use same image for both types of container isolation

Inventory move fix related to container groups

AlanCoding commented on awx/main/models/unified_jobs.py:
```diff
@@ -376,6 +394,8 @@ def create_unified_job(self, **kwargs):
             for fd, val in eager_fields.items():
                 setattr(unified_job, fd, val)
 
+        unified_job.execution_environment = self.resolve_execution_environment()
```
this seems fine to me, but with this addition, "execution_environment" should be removed from the list here.

https://github.com/ansible/awx/blob/4f1eafec26487892c58bbe49856c3e58755613d4/awx/main/models/jobs.py#L287

AlanCoding replied:
No, maybe I'm wrong, because you check

```python
        if self.execution_environment is not None:
            return self.execution_environment
```

so I guess the pattern is that `execution_environment` on the job can be set by the unified job template, and if it is, that comes highest precedence, so it is respected.

If going _this_ route, then it seems `execution_environment` should be added to the unified job fields for inventory sources, since those have historically allowed overriding at the UJT level too.

jbradberry replied:
> this seems fine to me, but with this addition, "execution_environment" should be removed from the list here.
> 
> https://github.com/ansible/awx/blob/4f1eafec26487892c58bbe49856c3e58755613d4/awx/main/models/jobs.py#L287

What's the reasoning behind that?

jbradberry replied:
> No, maybe I'm wrong, because you check
> 
> ```python
>         if self.execution_environment is not None:
>             return self.execution_environment
> ```
> 
> so I guess the pattern is that `execution_environment` on the job can be set by the unified job template, and if it is, that comes highest precedence, so it is respected.
> 
> If going _this_ route, then it seems `execution_environment` should be added to the unified job fields for inventory sources, since those have historically allowed overriding at the UJT level too.

The `execution_environment` field on a job (as opposed to a job template) is supposed to be the already precedence-resolved EE to be used while executing.

As I noted in another comment, `InventorySource` does have `execution_environment` defined on itself, by virtue of inheriting from `UnifiedJobTemplate`.

AlanCoding replied:
This is the list of fields for InventorySource to copy over when `create_unified_jobs` is called

https://github.com/ansible/awx/blob/4f1eafec26487892c58bbe49856c3e58755613d4/awx/main/models/inventory.py#L1030-L1034

This should contain `execution_environment` by the same measure that JT does, if that's the design.

jbradberry replied:
It made sense when I did it in the first PR, but I suspect that you are right that it is no longer needed then.

AlanCoding commented on awx/main/models/unified_jobs.py:
```diff
@@ -338,6 +339,23 @@ def notification_templates(self):
+            return self.organization.default_environment
+        if getattr(self, 'inventory', None) and self.inventory.organization is not None:
+            if self.inventory.organization.default_environment is not None:
+                return self.inventory.organization.default_environment
```
It seems strange that inventory would come after organization in the selection logic.

Also, that creates a lot of confusion with inventory updates, since there would be an EE on inventory and an EE on the inventory source. IIRC, with custom virtual environments, inventory sources could have a custom venv specified, but inventory could not.

jbradberry replied:
`Inventory` does not have an EE.  It is used here solely for the purpose of inferring an `Organization`, for jobs that fail to have one more directly defined.

`InventorySource` _does_ have an EE, however, I have been very clear about maintaining a strict separation between "EE as explicitly set for use when executing" (always `self.execution_environment`) and "EE as a fallback option" (always `self.default_environment`).

jbradberry replied:
To be more clear, `InventorySource` has `self.execution_environment`.  The only models that have `self.default_environment` are / will be `Organization` and `Project`.

AlanCoding replied:
Thanks, we're on the same page about inventory source.

But in what case is it desirable to use `inventory.organization.default_environment`? I don't think that's desirable to JTs, it would confuse more than it would help. I could see that for the case of inventory updates where the inventory source does not set an EE.

So maybe make this conditional specific to the inventory update job type?

jbradberry replied:
Is there a non-contrived case where a JT would not have an Org but would wind up getting its EE from an inventory Org instead?  And would using that EE actually be harmful instead of falling back all of the way to the global default EE?

AlanCoding commented on awx/main/models/credential/__init__.py:
```diff
@@ -1187,6 +1186,37 @@ def create(self):
+            'type': 'string',
+            'secret': True,
+        }, {
+            'id': 'token',
```
So this isn't hooked up yet, I'd like to get some definition around that. What I've seen before is

```
podman login -u='user' -p='token'
```

That's used for pushing. So obviously you need your token to push to quay.io. However, I have not experimented with pulling from quay private repos or anything like that. I'm also unclear how this will work with Automation Hub authentication, to use their image registry, and the details of that authentication are certain to be different. But this is just 1 credential type, so how would we deal with that? And does it even work?

Also, does this need some form of support on the ansible-runner side? I'm tempted to say that we're getting too far ahead of ourselves, and maybe we can just get through the upstream release without any support for pulling images that require authentication.

jbradberry replied:
The three registries I looked into all take either username+password or a token.  If someone has specific knowledge of what changes are needed here to do the right thing, I would love to hear it.
